pyk/src/pyk/kcfg/minimize.py

```diff
@@ -188,6 +200,94 @@ def _fold_lift(result: bool, finder_lifter: tuple[Callable, Callable]) -> bool:
             _fold_lift, [(self.kcfg.edges, self.lift_split_edge), (self.kcfg.splits, self.lift_split_split)], False
         )
 
+    def merge_nodes(self) -> bool:
+        """Merge targets of Split for cutting down the number of branches, using heuristics KCFGSemantics.is_mergeable.
+
+        Side Effect: The KCFG is rewritten by the following rewrite pattern,
+            - Match: A -|Split|-> A_i -|Edge|-> B_i
+            - Rewrite:
+                - if `B_x, B_y, ..., B_z are not mergeable` then unchanged
+                - if `B_x, B_y, ..., B_z are mergeable`, then
+                    - A -|Split|-> A_x or A_y or ... or A_z
+                    - A_x or A_y or ... or A_z -|Edge|-> B_x or B_y or ... or B_z
+                    - B_x or B_y or ... or B_z -|Split|-> B_x, B_y, ..., B_z
+
+        Specifically, when `B_merge = B_x or B_y or ... or B_z`
+        - `or`: fresh variables in places where the configurations differ
+        - `Edge` in A_merged -|Edge|-> B_merge: list of merged edges is from A_i -|Edge|-> B_i
+        - `Split` in B_merge -|Split|-> B_x, B_y, ..., B_z: subst for it is from A -|Split|-> A_1, A_2, ..., A_n
+        :param semantics: provides the is_mergeable heuristic
+        :return: whether any merge was performed
+        """
+
+        def _is_mergeable(x: KCFG.Edge | KCFG.MergedEdge, y: KCFG.Edge | KCFG.MergedEdge) -> bool:
+            return self.semantics.is_mergeable(x.target.cterm, y.target.cterm)
+
+        # ---- Match ----
+
+        # A -|Split|> Ai -|Edge/MergedEdge|> Mergeable Bi
+        sub_graphs: list[tuple[KCFG.Split, list[list[KCFG.Edge | KCFG.MergedEdge]]]] = []
+
+        for split in self.kcfg.splits():
+            _edges = [
+                single(self.kcfg.general_edges(source_id=ai))
+                for ai in split.target_ids
+                if self.kcfg.general_edges(source_id=ai)
+            ]
+            _partitions = partition(_edges, _is_mergeable)
+            if len(_partitions) < len(_edges):
+                sub_graphs.append((split, _partitions))
+
+        if not sub_graphs:
+            return False
+
+        # ---- Rewrite ----
+
+        for split, edge_partitions in sub_graphs:
+
+            # Remove the original sub-graphs
+            for p in edge_partitions:
+                if len(p) == 1:
+                    continue
+                for e in p:
+                    # TODO: remove the split and edges, then safely remove the nodes.
+                    self.kcfg.remove_node(e.source.id)
```

Comment:
I don't think this `TODO` is needed. Removing the node will go through and remove all the attached splits and edges as well: https://github.com/runtimeverification/k/blob/6e23a83e5380f9eae45a73e3956776f2e666da35/pyk/src/pyk/kcfg/kcfg.py#L764

Reply:
That's why I think it's just needed to remove the target nodes of the items of the split.

Reply:
Sure, this will remove all attached splits and edges. However, it may not only eliminate the matched splits and edges but could also inadvertently remove unexpected edges, such as those from other splits, covers, unbranches, and edges to the targets of the splits, as well as splits, covers, and unbranches from those targets.

I prefer not to remove all target nodes of the split because this would also eliminate partitions with a single element. These single-element partitions do not merge anything and should remain unchanged. To restore them after deletion, we need to retain the CTerms of those nodes and the information about the edges (Edge or MergedEdge). Therefore, it is simpler and clearer to keep them unchanged.